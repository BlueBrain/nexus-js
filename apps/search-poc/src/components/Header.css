.header-container {
  flex: 1 100%;
  display: flex;
  justify-content: space-between;
  margin: 0.5em 1em;
}

.header-container > .title {
  font-weight: lighter;
}

.header-container > .auth {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  flex-grow: 1;
  width: 3em;
  height: 3em;
}
