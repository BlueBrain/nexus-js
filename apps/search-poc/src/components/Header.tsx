import * as React from 'react';
import { Menu, Icon, Dropdown, Button } from 'antd';
import { UserManager, User } from 'oidc-client';
import { useNexusContext } from '@bbp/react-nexus';
import { Realm } from '@bbp/nexus-sdk';
import { SETTINGS } from '../config';
import { getConfig } from '../utils/auth';

import logoSvg from '../logo.svg';

const Login: React.FunctionComponent<{
  realms: Realm[];
  onRealmSelected(realm: Realm): void;
}> = ({ realms, onRealmSelected }) => {
  const menu = (
    <Menu>
      {realms.map(realm => (
        <Menu.Item key={realm['@id']} onClick={() => onRealmSelected(realm)}>
          {realm._label}
        </Menu.Item>
      ))}
    </Menu>
  );
  return (
    <Dropdown.Button overlay={menu} icon={<Icon type="login" />}>
      Login
    </Dropdown.Button>
  );
};
const Logout: React.FunctionComponent<{
  userName: string;
  onLogout(): void;
}> = ({ userName, onLogout }) => {
  return (
    <Button icon="logout" onClick={onLogout}>
      {userName}: Logout
    </Button>
  );
};

const Header: React.FunctionComponent<{
  userManager: UserManager | null;
  user: User | null;
}> = ({ userManager, user }) => {
  const nexus = useNexusContext();

  // fetch list of available realms when application mounts
  const [realms, setRealms] = React.useState<Realm[]>([]);
  React.useEffect(() => {
    nexus.Realm.list()
      .then(realms => {
        // nexus uses a realm called `serviceaccount` internally, we need to filter it out
        setRealms(
          realms._results.filter(
            realm => realm._label !== SETTINGS.serviceAccountName,
          ),
        );
      })
      .catch(e => console.error(e));
  }, [nexus]);

  const handleLogin = (realm: Realm) => {
    if (!userManager) {
      return;
    }
    const nextUserManagerConfig = getConfig(realm);
    const nextUserManager = new UserManager(nextUserManagerConfig);
    nextUserManager
      .signinRedirect()
      .then(() => {
        userManager.revokeAccessToken();
        userManager.clearStaleState();
        localStorage.setItem(
          SETTINGS.preferredRealmKey,
          JSON.stringify({ _issuer: nextUserManagerConfig.authority }),
        );
      })
      .catch(e => console.log(e.message));
  };

  const handleLogout = () => {
    if (!userManager) {
      return;
    }
    userManager
      .signoutRedirect()
      .then(() => {
        userManager.revokeAccessToken();
        userManager.clearStaleState();
        localStorage.removeItem(SETTINGS.preferredRealmKey);
      })
      .catch(e => console.log(e.message));
  };

  return (
    <div className="Header">
      <img className="Header_logo" src={logoSvg} alt="Nexus" />
      <div className="Header_auth">
        {user ? (
          <Logout onLogout={handleLogout} userName={user.profile.name} />
        ) : (
          <Login realms={realms} onRealmSelected={handleLogin} />
        )}
      </div>
    </div>
  );
};

export default Header;
