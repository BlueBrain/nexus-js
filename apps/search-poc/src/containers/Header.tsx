import * as React from 'react';
import { Menu, Icon, Dropdown, Button, notification } from 'antd';
import { UserManager } from 'oidc-client';
import { useNexusContext } from '@bbp/react-nexus';
import { Realm } from '@bbp/nexus-sdk';
import { SETTINGS } from '../config';
import { getConfig } from '../utils/auth';
import useUser from '../hooks/useUser';

const Login: React.FunctionComponent<{
  realms: Realm[];
  onRealmSelected(realm: Realm): void;
}> = ({ realms, onRealmSelected }) => {
  const [selectedRealm, setSelectedRealm] = React.useState<Realm>(realms[0]);

  React.useEffect(() => {
    if (!selectedRealm) {
      setSelectedRealm(realms[0]);
    }
  }, [realms, setSelectedRealm, selectedRealm]);

  const menu = (
    <Menu>
      {realms.map(realm => (
        <Menu.Item key={realm['@id']} onClick={() => setSelectedRealm(realm)}>
          {realm._label}
        </Menu.Item>
      ))}
    </Menu>
  );
  return selectedRealm ? (
    <Dropdown.Button
      overlay={menu}
      onClick={() => onRealmSelected(selectedRealm)}
      icon={<Icon type="down" />}
    >
      Login with {selectedRealm._label} <Icon type="login" />
    </Dropdown.Button>
  ) : null;
};
const Logout: React.FunctionComponent<{
  userName: string;
  onLogout(): void;
}> = ({ userName, onLogout }) => {
  return (
    <Button icon="logout" onClick={onLogout}>
      {userName}: Logout
    </Button>
  );
};

const Header: React.FunctionComponent = () => {
  const nexus = useNexusContext();
  const { userManager, user, error: userError } = useUser();
  // fetch list of available realms when application mounts
  const [realms, setRealms] = React.useState<Realm[]>([]);
  React.useEffect(() => {
    nexus.Realm.list()
      .then(realms => {
        // nexus uses a realm called `serviceaccount` internally, we need to filter it out
        setRealms(
          realms._results.filter(
            realm => realm._label !== SETTINGS.serviceAccountName,
          ),
        );
      })
      .catch(e => console.error(e));
  }, [nexus]);

  const handleLogin = (realm: Realm) => {
    if (!userManager) {
      return;
    }
    const nextUserManagerConfig = getConfig(realm);
    const nextUserManager = new UserManager(nextUserManagerConfig);
    nextUserManager
      .signinRedirect()
      .then(() => {
        userManager.revokeAccessToken();
        userManager.clearStaleState();
        localStorage.setItem(
          SETTINGS.preferredRealmKey,
          JSON.stringify({ _issuer: nextUserManagerConfig.authority }),
        );
      })
      .catch(e => console.log(e.message));
  };

  const handleLogout = () => {
    if (!userManager) {
      return;
    }
    userManager
      .signoutRedirect()
      .then(() => {
        userManager.revokeAccessToken();
        userManager.clearStaleState();
        localStorage.removeItem(SETTINGS.preferredRealmKey);
      })
      .catch(e => console.log(e.message));
  };

  if (userError) {
    notification.error({
      message: userError.name,
      description: userError.message,
    });
  }

  return (
    <div className="Header">
      <h1 className="title">
        <a href="/">BBP Studio</a>
      </h1>
      <div className="Header_auth">
        {user ? (
          <Logout onLogout={handleLogout} userName={user.profile.name} />
        ) : (
          <Login realms={realms} onRealmSelected={handleLogin} />
        )}
      </div>
    </div>
  );
};

export default Header;
