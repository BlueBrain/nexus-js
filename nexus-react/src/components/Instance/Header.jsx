import React from "react";
import FontAwesome from "react-fontawesome";
import CopyToClipboard from "../CopyToClipboard/index.jsx";
import style from "./instance.css";
import Type from "../Shapes/Type";

const Revision = rev => (
  <span style={{ float: "right", lineHeight: "2.25em" }}>
    <span>revision</span> <span>{rev}</span>
  </span>
);

const Name = instance => {
  let name =
    instance["skos:prefLabel"] ||
    instance["rdfs:label"] ||
    instance["schema:name"] ||
    instance["name"];
  let id = instance["@id"];
  if (!name) {
    let split = id.split("-");
    name = "..." + split[split.length - 1];
  }
  return (
    <div className="name">
      {name} <CopyToClipboard value={id} text="copy identifier" />
    </div>
  );
};

const Deprecated = deprecated => (
  <div className="deprecated">
    {deprecated ? (
      <small style={{ color: "#ff666680" }}>
        <FontAwesome name="exclamation" /> Deprecated
      </small>
    ) : null}
  </div>
);

const Header = instance => (
  <div className="instance-header">
    <div className="instance-link right">
      <CopyToClipboard
        value={location.href}
        icon="link"
        text="copy shareable link"
      />
    </div>
    <h1>{Name(instance)}</h1>
    <div className={style.instance}>
      {Revision(instance["nxv:rev"])}
      <Type type={instance["@type"]} />
      {Deprecated(instance["nxv:deprecated"])}
    </div>
  </div>
);
export default Header;
