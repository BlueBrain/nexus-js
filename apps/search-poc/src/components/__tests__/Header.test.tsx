import * as renderer from "react-test-renderer";
import Header from '../Header';
import React from "react";
import { shallow, mount } from "enzyme";


it("Header renders correctly", () => {
  const props = {
        handleLogin: jest.fn,
        handleLogout: jest.fn,
        userName: 'test',
        realms: [],
        Logout: () => {},
        Login: jest.fn()
    };

  const result = shallow(<Header {...props} />);
  expect(result).toMatchSnapshot();
});


