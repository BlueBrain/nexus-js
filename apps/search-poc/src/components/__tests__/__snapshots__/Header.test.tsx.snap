// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<div
  className="header-container"
>
  <h1
    className="title"
  >
    <a
      href="/"
    >
      BBP Studio
    </a>
  </h1>
  <div
    className="auth"
  >
    <Logout
      onLogout={[Function]}
      userName="test"
    />
  </div>
</div>
`;
